### PR TITLE
Handle WSL DXG teardown and missing VRAM segment info on Windows 10

### PR DESCRIPTION
**Pull request by hlky**
## Motivation

Windows 10 WSL2 does not currently work reliably with the `librocdxg` ROCm/HIP path. While Windows 11 / WDDM 3.1+ appears to be the intended supported environment, there are users on Windows 10 WSL2 systems where basic HIP device discovery and simple kernels can get close to working but currently fail on a few platform/runtime edges.

This PR is an MVP-style enablement patch for that Windows 10 WSL2 path. It addresses the two blockers observed during local testing:

- `VramAvail()` depends on WDDM segment type classification that is not available on Windows 10, causing bad `hipMemGetInfo()` behavior.
- Normal HIP process teardown can hang in the WSL DXG kernel path while destroying resident allocations.

The goal is not to declare Windows 10 WSL2 fully supported. The goal is to document and upstream the smallest changes that made basic ROCm/HIP workloads viable in this environment, and hopefully give AMD maintainers a concrete base to evaluate, refine, or replace with a more complete implementation.

## Technical Details

- `GpuMemory::~GpuMemory()` now evicts real resident allocations before freeing GPU VA / physical memory.
  - This gives normal user-mode teardown a chance to make the allocation non-resident before destruction.
  - The intent is to avoid hangs during WSL DXG allocation destruction.

- `WDDMDevice::GetSegmentId()` now accepts an optional `log_error` argument.
  - Existing callers keep the previous behavior by default.
  - `VramAvail()` uses the quiet path for the dGPU memory-segment lookup because missing segment classification can be an expected platform condition.

- `VramAvail()` now falls back to `LocalHeapSize()` for dGPU when the WDDM memory segment type is unavailable.
  - This avoids returning an error-like value through VRAM availability queries.
  - It provides a static fallback rather than live residency accounting.

Related discussion:

- https://github.com/ROCm/librocdxg/issues/22
- https://github.com/ROCm/librocdxg/issues/22#issuecomment-4323784066

## Test Plan

Tested in WSL2 with `HSA_ENABLE_DXG_DETECTION=1` and a locally rebuilt `librocdxg`.

Build:

```bash
cmake --build /root/rocdxg-build -j2
```

Runtime validation:

- `rocminfo` detects the GPU agent.
- Normal teardown probe using `hipGetDeviceCount()` exits cleanly.
- Vector-add HIP probe tested at:
  - 1 MiB + 4 bytes
  - 4 MiB
  - 64 MiB
- Repeated 4 MiB vector-add in 10 fresh processes without `wsl --shutdown`.
- `hipMemGetInfo()` before and after a 64 MiB allocation.
- Explicit `hipFree()` return checked in vector and allocation tests.

## Test Environment

- Host OS: Windows 10 IoT Enterprise LTSC 2021 / 21H2, build 19044
- WSL kernel: 6.6.114.1-microsoft-standard-WSL2
- GPU: AMD Radeon RX 9070 XT
- AMD driver: 32.0.23027.2005 / Adrenalin 26.2.2
- WSL distro: Ubuntu 24.04.4 LTS
- ROCm: 7.2.1
- ROCDXG package: rocdxg-roct 1.2.0
- librocdxg: /opt/rocm-7.2.1/lib/librocdxg.so.1.2.0

## Test Result

With this patch, basic HIP workloads became viable on the tested Windows 10 WSL2 system.

Passed:

- GPU discovery via `rocminfo`
- 1 MiB + 4 bytes vector-add
- 4 MiB vector-add
- 64 MiB vector-add
- 10 repeated fresh-process vector-add runs without `wsl --shutdown`
- `hipMemGetInfo()` before/after allocation
- explicit `hipFree()` path

Known limitation: abrupt process termination via `std::_Exit` can still bypass user-mode cleanup and leave teardown to the WSL/DXG kernel release path. This PR addresses normal ROCm/HIP cleanup, not forced process death.

`hipMemGetInfo()` now returns sane static VRAM values on Windows 10, but this fallback is not live residency accounting.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

## Disclosure

Codex assisted with the investigation, probing, and patch iteration. I reviewed the final diff and verified the test results before submitting.

### CODE REVIEW

File: src/wddm/gpu_memory.cpp [hunk 1]
```diff
@@ -46,6 +46,13 @@ GpuMemory::GpuMemory(WDDMDevice *device) : device_(device) {
 }
 
 GpuMemory::~GpuMemory() {
+  if (GpuAddress() != 0 &&
+      alloc_handles_ptr_ != nullptr && !(NumChunks() == 1 && *alloc_handles_ptr_ == 0) &&
+      !IsVirtual() && !IsPhysicalOnly()) {
+    // In WSL, destroying a still-resident allocation can block in
+    // dxgallocation_destroy while Hyper-V tears down the GPADL.
+    Evict();
```
Comment (fcui-amd) [inline src/wddm/gpu_memory.cpp]: For the destructor-side Evict() change, please check and log the Evict() return value instead of ignoring it. Also please confirm whether imported allocations are safe to evict from this path, and whether eviction needs to be synchronized with the paging fence before DestroyAllocation().

File: src/wddm/device.cpp [hunk 2]
```diff
@@ -170,8 +171,11 @@ uint64_t WDDMDevice::VramAvail(void) {
     return HSA_STATUS_ERROR;
 
   // local cpu-visible memory
-  if (!GetSegmentId(D3DKMT_QUERYSTATISTICS_SEGMENT_TYPE_MEMORY, segmentId))
+  if (!GetSegmentId(D3DKMT_QUERYSTATISTICS_SEGMENT_TYPE_MEMORY, segmentId, false)) {
+    if (IsDgpu())
```
Comment (fcui-amd) [inline src/wddm/device.cpp]: This does not get the actual available VRAM here, and it reports an incorrect value to upper-layer applications.

File: src/wddm/gpu_memory.cpp (hunk 1 above)
Comment (hlky) [reply src/wddm/gpu_memory.cpp]: Addressed in `22876f1`.

Changes made:

- `Evict()` return value is now checked and logged with `ErrorCodeToString()`.
- Added `GpuMemory::IsImported()` and exclude imported sysmem/vram allocations from this destructor-side eviction path.
- When destructor eviction is attempted, the code waits on the tracked paging fence before `FreePhysicalMemory()` / `DestroyAllocation()`.

One nuance: based on the existing comments elsewhere in this codebase, `Evict()` itself does not update the paging fence, so this waits on the runtime's tracked paging fence before destroy rather than claiming an eviction-specific fence value. Retested normal teardown and HIP vector/meminfo paths after the change.

File: src/wddm/device.cpp (hunk 2 above)
Comment (hlky) [reply src/wddm/device.cpp]: Agreed, the previous `LocalHeapSize()` fallback was too blunt.

Updated in `22876f1`: when the typed WDDM memory segment lookup is unavailable, `VramAvail()` now identifies the dGPU local VRAM segment from legacy segment properties (non-system, non-aperture, and `commit_limit == LocalHeapSize()`), then still queries `D3DKMTQueryStatistics` for `BytesResident` and returns `LocalHeapSize() - BytesResident`.

On the tested Windows 10 system this now produces a live residency-based value. Latest local check with a 64 MiB allocation:

```text
before=13425876992/17043521536
after=13358231552/17043521536
final=13425876992/17043521536
```

File: src/wddm/device.cpp [hunk 3]
```diff
@@ -193,8 +219,12 @@ uint64_t WDDMDevice::VramAvail(void) {
...
   }
 
-  if (IsDgpu())
-    return LocalHeapSize() - usedVis - usedInv;
+  if (IsDgpu()) {
+    uint64_t usedLocal = usedVis + usedInv;
+    if (usedLocal >= LocalHeapSize())
+      return 0;
+    return LocalHeapSize() - usedLocal;
+  }
```
Comment (fcui-amd) [inline src/wddm/device.cpp]: I don’t have a Windows 10 system available. Could you help verify whether this patch https://github.com/ROCm/librocdxg/tree/fcui/0521 works on your side?

File: src/wddm/gpu_memory.cpp (hunk 1 above)
Comment (fcui-amd) [reply src/wddm/gpu_memory.cpp]: Apologies — I missed that Evict() doesn't update the paging fence in our codebase when I asked for the fence wait.

File: src/wddm/device.cpp (hunk 3 above)
Comment (hlky) [reply src/wddm/device.cpp]: Thanks, I tested `fcui/0521` on Windows 10 WSL.

Environment:

- Host OS: Windows 10 IoT Enterprise LTSC 21H2, build 19044.7184
- GPU: Radeon RX 9070 XT / `gfx1201`
- Driver: 32.0.23027.2005 / Adrenalin 26.2.2
- WSL: 2.7.3.0, kernel 6.6.114.1-microsoft-standard-WSL2
- Distro: Ubuntu 24.04.4
- ROCm: 7.2.1

Positive result: after a clean `wsl --shutdown`, `hipMemGetInfo` behaved correctly in one full validation run. A 64 MiB allocation reduced free memory and `hipFree` restored it:

```text
meminfo alloc=67108864
before=14663172096/17043521536
after=14596063232/17043521536
final=14663172096/17043521536
hipFree=no error
```

The same run also passed 1 MiB + 4 bytes, 4 MiB, and 64 MiB vector-add tests.

However, the new segment-group VRAM query is not reliable on this Windows 10 setup. After another clean `wsl --shutdown`, before any prior HIP teardown hang in that WSL instance, the first vector test failed before producing probe stdout with:

```text
[VramAvail] Local segment group usage query failed (11)
[VramAvail] Local segment group usage query failed (11)
```

In this branch, `ErrorCode 11` is `InvalidParams`. The kernel log matched this as:

```text
dxgkio_query_statistics: Ioctl failed: -22
```

After that failure, the process wedged in the known DXG allocation cleanup path:

```text
vmbus_teardown_gpadl
dxgallocation_destroy
dxgkio_destroy_allocation
dxgk_ioctl
```

Comment (fcui-amd) [reply src/wddm/device.cpp]: Thanks for testing `fcui/0521` on Windows 10 and for the detailed repro — that’s very helpful.

You’re right: segment group usage query also requires WDDM 3.0+ (WDDM 3.1 on Win11). Relying on segment group query alone is not sufficient there.

The [fcui/0522](https://github.com/ROCm/librocdxg/tree/fcui/0522) branch restores the previous per-segment VramAvail() behavior and adds a segment-ID fallback: it still tries QuerySegmentInfo() / GetSegmentId() first, and if type lookup fails, it falls back to hardcoded segment IDs (with lazy segment enumeration only on the fallback path). On WDDM 3.0+ drivers, it continues to prefer segment group counters when available.

Could you give `fcui/0522` a try on your Windows 10 setup when you have a chance? I’m hoping this addresses the unreliable segment-group path while keeping live residency-based hipMemGetInfo() on Win10.

**Comment by fcui-amd** [comment]
and I suppose the Evict(); change you made could also guard with wddm version check:      Platform::instance().WddmVersion() < KMT_DRIVERVERSION_WDDM_3_0;

File: src/wddm/device.cpp (hunk 3 above)
Comment (hlky) [reply src/wddm/device.cpp]: Thanks, I tested `fcui/0522` on the same Windows 10 setup.

Branch tested: `946453d` (`wddm: use segment group counters for VramAvail with fallback`).

Good news: this branch appears to address the `fcui/0521` VRAM reporting issue for us. I no longer see the `VramAvail` segment-group `InvalidParams` / `dxgkio_query_statistics: -22` failure.

A 64 MiB `hipMemGetInfo` probe completed successfully:

```text
devices: 1
device0: AMD Radeon RX 9070 XT gcn=gfx1201 totalGlobalMem=17043521536
meminfo alloc=67108864
before=15212036096/17043521536
after=15144390656/17043521536
final=15212036096/17043521536
hipFree=no error
```

I also repeated the 64 MiB `hipMemGetInfo` probe across three fresh `wsl --shutdown` starts. All three runs succeeded, stderr was empty, and I did not see the intermittent `VramAvail` / `dxgkio_query_statistics: -22` failure that occurred with `fcui/0521`.

```text
before=15209132032 after=15140954112 final=15209132032
before=15226454016 after=15158808576 final=15226454016
before=15214972928 after=15147327488 final=15214972928
```

A tiny compute probe also completed correctly:

```text
devices: 1
result: 42.0
marker: after_compute
```

And the explicit `hipFree` path reached:

```text
marker: after_free
marker: before_return
```

However, the normal process teardown hang is still present. Even a `hipGetDeviceCount`-only probe reaches `before_return` and then remains stuck in:

```text
vmbus_teardown_gpadl
dxgallocation_destroy
dxgkio_destroy_allocation
dxgk_ioctl
```

So my read is:

- `fcui/0522` appears to fix/improve the Win10 `VramAvail()` fallback behavior.
- It does not address the separate DXG allocation teardown hang.
- The cleanup fix from this PR still seems needed independently.

Comment (fcui-amd) [reply src/wddm/device.cpp]: Thanks a lot for re-testing on Win10 and for the detailed logs, this is very helpful.

My understanding is:
- fcui/0522 fixes the Win10 VramAvail() path.
- The teardown hang appears to be a separate issue.

btw, wasn’t the destructor-side Evict() cleanup change in your PR intended to address this teardown hang path?

Comment (hlky) [reply src/wddm/device.cpp]: Yes, exactly. The destructor-side `Evict()` change in this PR is intended to address the normal user-mode teardown path.

What we observed before this PR was that ordinary process exit could leave resident allocations to be destroyed by the DXG process-release path, where WSL then hung in:

```text
vmbus_teardown_gpadl
dxgallocation_destroy
dxgprocess_release
```

This PR changes `GpuMemory::~GpuMemory()` so owned resident allocations are explicitly evicted before `DestroyAllocation`, and it waits the paging fence before freeing/destroying them. In our testing, that fixed the normal return / `hipFree` / `hipDeviceReset` cleanup cases.

The remaining caveat is forced teardown. `_Exit`, crash-like exit, or SIGKILL bypass user-mode destructors, so the destructor-side cleanup does not run. Those cases can still hit the kernel/DXG process-release path directly, and we still see hangs there.

So I’d separate it this way:

- `fcui/0522`: fixes the Win10 `VramAvail()` reporting path.
- This PR’s `GpuMemory::~GpuMemory()` eviction/fence cleanup: fixes normal user-mode teardown.
- Forced `_Exit` / SIGKILL-style teardown: still appears to require a lower-level DXG/driver/kernel-side fix, since user-mode destructors are bypassed.

Comment (fcui-amd) [reply src/wddm/device.cpp]: https://github.com/ROCm/librocdxg/pull/41 is merged. could you guard the Evict() with Platform::instance().WddmVersion() < KMT_DRIVERVERSION_WDDM_3_0 ?

Comment (hlky) [reply src/wddm/device.cpp]: Done, thanks. I rebased this PR onto current `develop` after #41, dropped the `src/wddm/device.cpp` / `VramAvail()` changes from this PR, and narrowed the remaining change to the `GpuMemory` teardown path.

`Evict()` is now guarded with:

```cpp
Platform::instance().WddmVersion() < KMT_DRIVERVERSION_WDDM_3_0
```

so this cleanup path only applies on older WDDM versions.

I rebuilt and retested on the same Windows 10 setup. The rebased branch now passes the normal teardown cases that previously hung:

- `hipGetDeviceCount`-only probe exits cleanly
- HIP compute probe exits cleanly and returns `42.0`
- explicit `hipFree` path reaches `before_return` and exits cleanly
- 64 MiB `hipMemGetInfo` alloc/free probe exits cleanly
- `rocminfo` prints `*** Done ***` and exits with `rc=0`

I also did an A/B check with `rocminfo` against the packaged/current release library on the same setup. The release path still prints `*** Done ***`, but then remains stuck in:

```text
vmbus_teardown_gpadl
dxgallocation_destroy
dxgkio_destroy_allocation
dxgk_ioctl
```

So this also appears to address the Win10 `rocminfo` hang described in #17 for the normal process teardown path.

The known remaining limitation is forced process teardown, e.g. `_Exit` / crash / SIGKILL-style cleanup, because that bypasses librocdxg user-mode destructors. In that case cleanup is left to the WSL `/dev/dxg` kernel driver process-release path, where we still see the hang.

I would still be hesitant to claim that every ROCm workload is now supported on Win10 WSL. What this PR appears to fix is the normal user-mode allocation teardown hang we can reproduce with small HIP probes and `rocminfo`. Broader Win10 WSL support would still need AMD’s judgment and more coverage.

Happy to run additional targeted tests on this Win10 setup if there are specific workloads or edge cases you want checked.

**Review by fcui-amd — APPROVED** [review]
